api/argo-templates/ncn/worker.drain.yaml

```diff
@@ -22,36 +22,108 @@
-    # * OnWorkflowCompletion - delete pods when workflow is completed
-    # * OnWorkflowSuccess - delete pods when workflow is successful
-    strategy: OnPodCompletion
-  affinity:
-    nodeAffinity:
-      requiredDuringSchedulingIgnoredDuringExecution:
-        nodeSelectorTerms:
-        - matchExpressions:
-          - key: kubernetes.io/hostname
-            operator: NotIn
-            values:
-            {{- range $index,$value := .TargetNcns }}
-            - {{$value -}}
-            {{- end }}
-  arguments:
-    parameters:
-    - name: scriptContent
-      value: |
-        sleep $((RANDOM % 10))
-    - name: dryRun
-      value: {{"{{inputs.parameters.dryRun}}"}}
-  workflowTemplateRef:
-    name: {{"{{inputs.parameters.workflowtemplate}}"}}
+tasks:
+  - name: wait-for-cfs
+    templateRef:
+      name: ssh-with-kubeconfig
+      template: bash-script-example
+    arguments:
+      parameters:
+        - name: dryRun
+          value: "{{ `{{inputs.parameters.dryRun}}` }}"
+        - name: scriptContent
+          value: |
+            /usr/share/doc/csm/upgrade/1.2/scripts/cfs/wait_for_configuration.sh \
+              --xnames \
+              $(ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null {{ `{{inputs.parameters.targetNcn}}` }} 'cat /etc/cray/xname')
+  - name: ensure-nexus-can-start-on-any-node
+    templateRef:
+      name: ssh-with-kubeconfig
+      template: bash-script-example
+    arguments:
+      parameters:
+        - name: dryRun
+          value: "{{ `{{inputs.parameters.dryRun}}` }}"
+        - name: scriptContent
+          value: |
+            workers="$(kubectl get node --selector='!node-role.kubernetes.io/master' -o name | sed -e 's,^node/,,' | paste -sd,)"
+            export PDSH_SSH_ARGS_APPEND="-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
+            kubectl get configmap -n nexus cray-precache-images -o json | jq -r '.data.images_to_cache' | while read image; do echo >&2 "+ caching $image"; pdsh -w "$workers" "crictl pull $image" 2>/dev/null; done
+  - name: ensure-etcd-pods
+    templateRef:
+      name: ssh-with-kubeconfig
+      template: bash-script-example
+    arguments:
+      parameters:
+        - name: dryRun
+          value: "{{ `{{inputs.parameters.dryRun}}` }}"
+        - name: scriptContent
+          value: |
+            while [[ "$(kubectl get po -A -l 'app=etcd' | grep -v "Running"| wc -l)" != "1" ]]; do
+                echo "Some etcd pods are not in running state, wait for 5s ..."
+                kubectl get po -A -l 'app=etcd' | grep -v "Running"
+                sleep 5
+            done
+            etcdClusters=$(kubectl get Etcdclusters -n services | grep "cray-"|awk '{print $1}')
+            for cluster in $etcdClusters
+            do
+                numOfPods=$(kubectl get pods -A -l 'app=etcd'| grep $cluster | grep "Running" | wc -l)
+                if [[ $numOfPods -ne 3 ]];then
```

Comment:
Where does the magic number of 3 come from? Can customers change the number/count of etcd systems to 5 or 7?